internal/services/network/network_manager_commit_resource.go

```diff
@@ -0,0 +1,342 @@
+package network
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/location"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/azure"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/network/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/network/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+	"github.com/tombuildsstuff/kermit/sdk/network/2022-07-01/network"
+)
+
+type ManagerCommitModel struct {
+	NetworkManagerId string   `tfschema:"network_manager_id"`
+	ScopeAccess      string   `tfschema:"scope_access"`
+	Location         string   `tfschema:"location"`
+	ConfigurationIds []string `tfschema:"configuration_ids"`
+	DeploymentStatus string   `tfschema:"deployment_status"`
+}
+
+type ManagerCommitResource struct{}
+
+func (r ManagerCommitResource) ResourceType() string {
+	return "azurerm_network_manager_commit"
+}
+
+func (r ManagerCommitResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return validate.NetworkManagerCommitID
+}
+
+func (r ManagerCommitResource) ModelObject() interface{} {
+	return &ManagerCommitModel{}
+}
+
+func (r ManagerCommitResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"network_manager_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ValidateFunc: validate.NetworkManagerID,
+		},
+
+		"location": commonschema.Location(),
+
+		"scope_access": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				string(network.ConfigurationTypeConnectivity),
+				string(network.ConfigurationTypeSecurityAdmin),
+			}, false),
+		},
+
+		"configuration_ids": {
+			Type:     pluginsdk.TypeList,
+			Required: true,
+			Elem: &pluginsdk.Schema{
+				Type:         pluginsdk.TypeString,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```

Comment:
since this resource is 1:1 with the Network Manager, this should be embedded within the Network Manager resource as blocks rather than being a separate resource - presumably with one for `connectivity` and one for `security_admin` - which also removes the need for the flag within the `features` block, can we update this?

Reply:
A Network Manager can have more than one commit. Actually the commit is used to deploy specfic type of configuration to specific location follow a [goal state model](https://learn.microsoft.com/en-us/azure/virtual-network-manager/concept-deployments#-goal-state-model), so we design the ID format to be like `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourceGroup1/providers/Microsoft.Network/networkManagers/networkManager1/commit|eastus|Connectivity`. From Portal:
![image](https://user-images.githubusercontent.com/104055472/218704482-d6e37ebd-4c82-48ab-85ed-cb07d69cb0f1.png)

Reply:
wouldn't that just mean you allow multiple commit blocks in the network manager resource then?

Reply:
The dependency is that Commit depends on the Configurations, Configurations depends on Network Manager, and a Commit can contains more than one Configurations. An example is like:
```hcl
resource "azurerm_network_manager" "test" {
  name                = "acctest-nm"
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_network_manager_connectivity_configuration" "test" {
  name                  = "acctest-nmcc"
  network_manager_id    = azurerm_network_manager.test.id
}

resource "azurerm_network_manager_commit" "test" {
  network_manager_id = azurerm_network_manager.test.id
  location           = "eastus"
  scope_access       = "Connectivity"
  configuration_ids  = [azurerm_network_manager_connectivity_configuration.test.id]
}
```

Allowing commits as blocks in the Network Manager will lead to cycle import.

Reply:
@teowa per the link above, it appears that the Goal State Model is contrary to how Terraform Resources are expected to work - since at the end of the `terraform apply` the resource has to be in a fully provisioned/stable state - so the question becomes how does this resource make sense in Terraform?

Reply:
@tombuildsstuff  Does it make sense if we regard the Goal State as the resource Terraform manages, and how to reach to Goal State is done in the service backend. In TF we can create/update/read/delete the Goal State, and we can add logics to wait until the deployment finished so it will finally reach a stable state. 

And the special case is that when a Configuration is updated and if user want to update the deployment, they need to use [`life_cycle.replace_triggered_by`](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#replace_triggered_by) to refresh the Goal State, and the new feature toggle `manager_commit_keep_on_destroy`  is to skip delete step while replacing, to prevent down time or releated security issue.
```hcl
provider "azurerm" {
  features {
    network {
      manager_commit_keep_on_destroy = true
    }
  }
}

resource "azurerm_network_manager_commit" "test" {
  network_manager_id = azurerm_network_manager.test.id
  location           = "eastus"
  scope_access       = "SecurityAdmin"
  configuration_ids  = [azurerm_network_manager_security_admin_configuration.test.id]
  depends_on         = [azurerm_network_manager_admin_rule.test]
  lifecycle {
    replace_triggered_by = [
      azurerm_network_manager_security_admin_configuration.test,
      azurerm_network_manager_admin_rule_collection.test,
      azurerm_network_manager_admin_rule.test,
    ]
  }
}
```

Reply:
@teowa we definitely shouldn't be exposing a `feature` flag to not delete resources, when Terraform destroys a Resource it should be entirely removed - if Terraform's not doing what it says it'll do (deleting the resource), then how can users rely on Terraform?

Rather than deleting the resource and requiring users to workaround the issue, we could instead handle this during the Update function - **however** the question remains: how does this resource make sense in Terraform?

Given that the Network Manager Commit resource is making configuration changes to other resources (for example, [Network Security Rules](https://learn.microsoft.com/en-us/azure/virtual-network-manager/concept-network-manager-scope)), if the parent resources are managed by Terraform they'll get flagged for removal during the next apply? As such regardless of _how_ we workaround the _Goal State_ issue, I believe this'd cause drift on other resources, and issues at plan/apply time in the future?

Perhaps I'm missing some additional context, but based on what I'm seeing in the documentation I don't think there's a way for us to ship this resource that isn't going to cause us issues in the medium-term (particularly around the drift to other resources and `terraform-plugin-framework` / protocol v6)- as such I believe the design of this Resource is entirely at odds with how Terraform expects resources to work?

Reply:
@tombuildsstuff thanks for the insights. For the `feature` flag of not deleting the resources, agree that this breaks Terraform consistence and maybe we can add a property like `update_with` to link to other resource to trigger a update. 

For the drift to other resources, the documents is a little confusing, the mentioned Network Security Rules should be `azurerm_network_manager_admin_rule` resource, and Network Manager Commit is to make the existing `azurerm_network_manager_admin_rule` applied to certain scope (controlled by `azurerm_network_manager.scopes` which is a subscription or management group). We have confirmed with service team the Network Manager Commit will not make any changes to other [`azurerm_network_security_rule`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule) resources.

But when Network Manager Commit is deploying the `azurerm_network_manager_connectivity_configuration` resource, at service backend it might modify or deleting existing `azurerm_virtual_network_peering` resource to reach the `hub and spoke` topology. But by design as the `azurerm_network_manger` and its sub-resource `azurerm_network_manager_connectivity_configuration` is to control the network in a big scope like a Subscription or Management Group, seems like other policy related resource. I think we can add in docs tell users do not use single network peering together with Network Manager.

Reply:
> So this will be changing Network Security Rules within a Network Security Group, which'll be an issue?

No, Network Manager will not make changes to Network Security Rules within a Network Security Group, it works by its own `azurerm_network_manager_admin_rule` resources.

---
> so I don't think that's going to be something that we can ship since it'll break things in the future when we update the protocol/Terraform SDK that's being used.

Sorry, I do not quite understand why updating the protocol/Terraform SDK will break things. Could you elaborate more on this ? By my understanding, breaking things may happen only when  `azurerm_virtual_network_peering`  and Network Manager are configured simultaneously. But we can tell users by doc not to use `azurerm_virtual_network_peering`  and Network Manager at the same time to avoid this. This is just like Subnets defined in-line within the [Virtual Network resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_network) and a standalone [Subnet resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/subnet), as well as the [KeyVault Access Policy](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault_access_policy) can be both inline or standalone. And the [Auto Scale Setting](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_autoscale_setting) will change the instance number of Scale Set resource. WDYT?

Reply:
@katbyte , from the [doc](https://learn.microsoft.com/en-us/azure/virtual-network-manager/concept-security-admins#the-order-of-evaluation), Security admin rules are evaluated before network security rules. And I have also confirmed with service team that Network Manager will NOT touch NSGs. They are in different layers. 

As for possible conflict with peering resource, can we add[notes](https://github.com/hashicorp/terraform-provider-azurerm/pull/20451/files#diff-f2ef843dfaa2763b4af2d7e5fc9dc9bc9a3b0e6e1bfd01ab012ff2394ea007d5R13) in the provider docs, if users don't config Network Peering and Network Manger Commit together, (they can also do this if they know how the Network Manager [scope](https://learn.microsoft.com/en-us/azure/virtual-network-manager/concept-network-manager-scope#scope-applicability) works), conflict can be avoid in their config file. Per above, I think no `computed` or `ignore_changes` is need to be added. Is this OK?

Reply:
from the docs there, it seems like it starts an "eventually consistent"  deployment. how do we, terraform, know when that is completed/done given there is nothign to poll on an ensure the create finished at the same time all these deployments finish?

Reply:
👍  thanks for the details

Reply:
please continue to review this, thanks.